### PR TITLE
Library Hubs and Music Stations

### PR DESCRIPTION
**Pull request by blacktwin**
Adding Station support as requested in #361

Adding Library Hubs. Returns "More By {ARTIST}" or "Rediscover"

- [x] New feature (non-breaking change which adds functionality)

### CODE REVIEW

**Comment by blacktwin** [comment]
Code snippets:
```python
for y in plex.library.section('Music').filterFields(mediaType='album'):
    print(vars(y))
    break
"""
{   '_data': <Element 'Field' at 0x0000022F716CF638>,
    '_details_key': '',
    '_initpath': '/library/sections/27/all?type=9',
    '_server': <PlexServer:http://127.0.0.1:324>,
    'key': 'album.title',
    'operators': [   <Operator:=:contains>,
                     <Operator:!=:does-not-contain>,
                     <Operator:==:is>,
                     <Operator:!==:is-not>,
                     <Operator:<=:begins-with>,
                     <Operator:>=:ends-with>],
    'subType': None,
    'title': 'Album Title',
    'type': 'string'}
"""
for sort in plex.library.section('TV Shows')._sorts():
    print(vars(sort))
    break
"""
{   '_data': <Element 'Directory' at 0x0000022F716C73B8>,
    '_details_key': '',
    '_initpath': '/library/sections/3/sorts',
    '_server': <PlexServer:http://127.0.0.1:324>,
    'defaultDirection': 'asc',
    'descKey': 'titleSort:desc',
    'firstCharacterKey': '/library/sections/3/firstCharacter',
    'key': 'titleSort',
    'title': 'Title'}
"""
for first in plex.library.section('Movies').firstCharacter():
    print(vars(first))
    break
"""
{   '_data': <Element 'Directory' at 0x0000022F716D69A8>,
    '_details_key': '',
    '_initpath': '/library/sections/2/firstCharacter',
    '_server': <PlexServer:http://127.0.0.1:324>,
    'key': '%23',
    'size': '16',
    'title': '#'}
"""
```

**Comment by blacktwin** [comment]
With the removal of the hardcoded `ALLOWED_FILTERS` `ALLOWED_SORT` and `BOOLEAN_FILTER` searching and syncing may take longer and cause an additional query but the upside should be that these lists are now dynamic.

File: plexapi/library.py [hunk 1]
```diff
@@ -455,6 +449,49 @@ def all(self, sort=None, **kwargs):
...
+            items.append(sort)
+        return items
+
+    def filterFields(self, mediaType=None):
+        items = []
+        key = '/library/sections/%s/filters?includeMeta=1' % self.key
+        data = self._server.query(key)
+        for meta in data.iter('Meta'):
+            for metaType in meta.iter('Type'):
+                if mediaType and metaType.attrib.get('type') == mediaType:
```
Comment (Hellowlol) [inline plexapi/library.py:477]: simplify this one.

if not mediatype or ...

File: plexapi/library.py [hunk 2]
```diff
@@ -1083,6 +1111,111 @@ def __len__(self):
...
+        """ Load attribute values from Plex XML response. """
+        self._data = data
+        self.defaultDirection = data.attrib.get('defaultDirection')
+        self.descKey = data.attrib.get('descKey')
+        self.key = data.attrib.get('key')
+        self.title = data.attrib.get('title')
+        self.firstCharacterKey = data.attrib.get('firstCharacterKey')
+
+
+class Field(PlexObject):
```
Comment (Hellowlol) [inline plexapi/library.py:1163]: We already have a class called Field. Maybe change the name or atleast make a better docstring.

Comment (blacktwin) [reply plexapi/library.py:1163]: oh yeah docstrings need updated. I based the name off the xml tagging. Technically they can both exist after 3746b4e . But, yeah, I can change it.

File: plexapi/library.py (hunk 1 above)
Comment (blacktwin) [reply plexapi/library.py:477]: Yeah, that one was tripping me up. I'm repeating the same for loop for both but I get the mediaType isolated when called. I'll move this PR to draft while I figure it out. I'm open to suggestions, of course.

Comment (Hellowlol) [reply plexapi/library.py:477]: What your doing is fine, but just use the or so you check both conditions at the same time.

**Comment by blacktwin** [comment]
folders method is kinda neat.

```python
for item in plex.library.section('Music').folders():
    for sub in item.subfolders():
        for track in sub.subfolders():
            print(vars(track))
```
We can get to the media item through the `/folder` endpoint.

**Comment by blacktwin** [comment]
With the exception of the decrease in coverage, this PR is gtg.

**Pull request merged**